src/v/pandaproxy/schema_registry/util.h

```diff
@@ -47,4 +54,44 @@ result<schema_definition> make_schema_definition(std::string_view sv) {
       ss::sstring{str_buf.GetString(), str_buf.GetSize()}};
 }
 
+///\brief The first invocation of one_shot::operator()() will invoke func and
+/// wait for it to finish. Concurrent invocatons will also wait.
+///
+/// On success, all waiters will be allowed to continue. Successive invocations
+/// of one_shot::operator()() will return ss::now().
+///
+/// If func fails, waiters will receive the error, and one_shot will be reset.
+/// Successive calls to operator()() will restart the process.
+class one_shot {
+    enum class state { empty, started, available };
+    using futurator = ss::futurize<ss::semaphore_units<>>;
+
+public:
+    explicit one_shot(ss::noncopyable_function<ss::future<>()> func)
+      : _func{std::move(func)} {}
+    futurator::type operator()() {
+        if (likely(_started_sem.available_units() != 0)) {
+            return ss::get_units(_started_sem, 1);
+        }
+        auto units = ss::consume_units(_started_sem, 1);
+        return _func().then_wrapped(
+          [this, units{std::move(units)}](ss::future<> f) mutable noexcept {
+              if (f.failed()) {
+                  units.release();
+                  auto ex = f.get_exception();
+                  _started_sem.broken(ex);
+                  _started_sem = ss::semaphore{0};
+                  return futurator::make_exception_future(ex);
+              }
+
+              _started_sem.signal(_started_sem.max_counter());
+              return futurator::convert(std::move(units));
+          });
```

Comment:
looks correct at first glance, but it mixes several semaphore apis so I'd like to take a closer look tomorrow.

I do think that it can be much simpler by representing the states explicitly. something like:

```cpp
bool initialized = false;
ss::semaphore sem{1};
ss::noncopyable_function<ss::future<>()> func;

return ss::with_semaphore(sem, 1, [this] {
   if (likely(initialized)) {
      return ss::now();
   }
   return func.then([this] {
      initialization = false;
      sem.signal(ss::semaphore::max_counter() - 1);
   });
});
```